clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1138,7 +1138,18 @@ private Call makeMetadataCall(long now) {
                 @Override
                 public void handleResponse(AbstractResponse abstractResponse) {
                     MetadataResponse response = (MetadataResponse) abstractResponse;
-                    metadataManager.update(response.cluster(), time.milliseconds());
+                    long now = time.milliseconds();
+                    metadataManager.update(response.cluster(), now);
```

Comment:
This is a good find. What happens if the metadata request itself is queued up to be sent to a node which is no longer online? Will it be stuck in `callsToSend` until it times out? I am wondering if we should check `NetworkClient.connectionFailed` after every poll() for all pending requests and reenqueue them as we are doing here. This is what we do in `ConsumerNetworkClient`.

Reply:
@hachikuji Thanks for the review. I think metadata requests are handled differently because they use `LeastLoadedNodeProvider` which assigns a node only if a ready node is available. If no ready node is available, then the call stays in `pendingRequests` and gets moved to `callsToSend` only when a node becomes ready after a subsequent poll. If a ready node is available, then it gets moved to `callsToSend` and is removed from `callsToSend` in the same iteration when the request is queued for send. If disconnection is processed in a subsequent poll, then the call is failed and retried using the retry path. We would never expect to see a metadata request in `callsToSend` at this point. Does that make sense?

Reply:
Hmm, I'm not sure about that. The metadata request uses `MetadataUpdateNodeIdProvider`, which just calls `leastLoadedNode` directly. But the node chosen may not have an established connection, or am I missing something?